File: src/_02_StructuralPattern/_02_06_FlyWeight/FontFactory.js
import Font from "./Font.js";

export default class FontFactory {
  /** @type { Map<string, Font> } */
  #cache = new Map();

  /**
   * @param { string } font 
   * @returns { Font }
   */
  getFont(font) {
    if(this.#cache.has(font)) {
      return this.#cache.get(font);
    } else {
      const fontInfo = font.split(":");
      const newFont = new Font(fontInfo[0], +fontInfo[1]);
      this.#cache.set(font, newFont);

      return newFont;
    }
  }
}

File: src/_02_StructuralPattern/_02_06_FlyWeight/java/after/FontFactory.java
package _02_StructuralPattern._02_06_FlyWeight.java.after;

import java.util.HashMap;
import java.util.Map;

public class FontFactory {
  
  private Map<String, Font> cache = new HashMap<String, Font>();

  public Font getFont(String font) {
    if(cache.containsKey(font)) {
      return cache.get(font);
    } else {
      String[] fontInfo = font.split(":");
      Font newFont = new Font(fontInfo[0], Integer.parseInt(fontInfo[1]));
      cache.put(font, newFont);

      return newFont;
    }
  }
  
}
